coreutils-6.10/obj-llvm/man/tr.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.35.
.TH TR "1" "June 2016" "GNU coreutils 6.10-dirty" "User Commands"
.SH NAME
tr \- translate or delete characters
.SH SYNOPSIS
.B tr
[\fIOPTION\fR]... \fISET1 \fR[\fISET2\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Translate, squeeze, and/or delete characters from standard input,
writing to standard output.
.TP
\fB\-c\fR, \fB\-C\fR, \fB\-\-complement\fR
first complement SET1
.TP
\fB\-d\fR, \fB\-\-delete\fR
delete characters in SET1, do not translate
.TP
\fB\-s\fR, \fB\-\-squeeze\-repeats\fR
replace each input sequence of a repeated character
that is listed in SET1 with a single occurrence
of that character
.TP
\fB\-t\fR, \fB\-\-truncate\-set1\fR
first truncate SET1 to length of SET2
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
SETs are specified as strings of characters.  Most represent themselves.
Interpreted sequences are:
.TP
\eNNN
character with octal value NNN (1 to 3 octal digits)
.TP
\e\e
backslash
.TP
\ea
audible BEL
.TP
\eb
backspace
.TP
\ef
form feed
.TP
\en
new line
.TP
\er
return
.TP
\et
horizontal tab
.TP
\ev
vertical tab
.TP
CHAR1\-CHAR2
all characters from CHAR1 to CHAR2 in ascending order
.TP
[CHAR*]
in SET2, copies of CHAR until length of SET1
.TP
[CHAR*REPEAT]
REPEAT copies of CHAR, REPEAT octal if starting with 0
.TP
[:alnum:]
all letters and digits
.TP
[:alpha:]
all letters
.TP
[:blank:]
all horizontal whitespace
.TP
[:cntrl:]
all control characters
.TP
[:digit:]
all digits
.TP
[:graph:]
all printable characters, not including space
.TP
[:lower:]
all lower case letters
.TP
[:print:]
all printable characters, including space
.TP
[:punct:]
all punctuation characters
.TP
[:space:]
all horizontal or vertical whitespace
.TP
[:upper:]
all upper case letters
.TP
[:xdigit:]
all hexadecimal digits
.TP
[=CHAR=]
all characters which are equivalent to CHAR
.PP
Translation occurs if \fB\-d\fR is not given and both SET1 and SET2 appear.
\fB\-t\fR may be used only when translating.  SET2 is extended to length of
SET1 by repeating its last character as necessary.  Excess characters
of SET2 are ignored.  Only [:lower:] and [:upper:] are guaranteed to
expand in ascending order; used in SET2 while translating, they may
only be used in pairs to specify case conversion.  \fB\-s\fR uses SET1 if not
translating nor deleting; else squeezing uses SET2 and occurs after
translation or deletion.
.SH AUTHOR
Written by Jim Meyering.
.SH "REPORTING BUGS"
Report bugs to <bug\-coreutils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2008 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B tr
is maintained as a Texinfo manual.  If the
.B info
and
.B tr
programs are properly installed at your site, the command
.IP
.B info tr
.PP
should give you access to the complete manual.


basic/tr.c
/*
   tr - translate characters            Author: Michiel Huisjes 
   Usage: tr [-cds] [string1 [string2]]
      c: take complement of string1
      d: delete input characters coded string1
      s: squeeze multiple output characters of string2 into one character

clang -emit-llvm -c -g tr.c -o tr.bc
time ~/git/tracerx/klee/Release+Asserts/bin/klee -libc=uclibc --posix-runtime -only-output-states-covering-new -allow-external-sym-calls ./tr.bc --sym-arg 1 --sym-arg 1 --sym-arg 1 --sym-files 2 2000 --max-fail 1 

clang -emit-llvm -c -g tr.c -o tr.bc
time ~/git/original/klee/Release+Asserts/bin/klee -libc=uclibc --posix-runtime -only-output-states-covering-new -allow-external-sym-calls ./tr.bc --sym-arg 1 --sym-arg 1 --sym-arg 1 --sym-files 2 2000 --max-fail 1 


*/ 
 
 #define BUFFER_SIZE     1024
 #define ASCII           0377
 
 typedef char BOOL;
 #define TRUE    1
 #define FALSE   0
 
 #define NIL_PTR         ((char *) 0)
 
 BOOL com_fl, del_fl, sq_fl;
 
 unsigned char output[BUFFER_SIZE], input[BUFFER_SIZE];
 unsigned char vector[ASCII + 1];
 BOOL invec[ASCII + 1], outvec[ASCII + 1];
 
 short in_index, out_index;
 
 #include <sys/types.h>
 #include <string.h>
 #include <stdlib.h>
 #include <unistd.h>
 
/*
 _PROTOTYPE(int main, (int argc, char **argv));
 _PROTOTYPE(void convert, (void));
 _PROTOTYPE(void map, (unsigned char *string1, unsigned char *string2));
 _PROTOTYPE(void expand, (char *arg, unsigned char *buffer));
 _PROTOTYPE(void complement, (unsigned char *buffer));*/

int main (int argc, char **argv);
void convert(void);
void map(unsigned char *string1, unsigned char *string2);
void expand(char *arg, unsigned char *buffer);
void complement(unsigned char *buffer);
 
int main(argc, argv)
int argc;
char *argv[];
//int main()
 { 
   //int SIZE = 3;
   //int argc;
   //char *argv[SIZE];
   //klee_make_symbolic(argv, sizeof argv, "argv");
   //argc = SIZE;
 
   register unsigned char *ptr;
   int index = 1;
   short i;
 
   if (argc > 1 && argv[index][0] == '-') {
         for (ptr = (unsigned char *) &argv[index][1]; *ptr; ptr++) {
                 switch (*ptr) {
                     case 'c':   com_fl = TRUE;  break;
                     case 'd':   del_fl = TRUE;  break;
                     case 's':   sq_fl = TRUE;   break;
                     default:
                         write(2,"Usage: tr [-cds] [string1 [string2]].\n", 38);
                         exit(1);
                 }
         }
         index++;
   }
   for (i = 0; i <= ASCII; i++) {
         vector[i] = i;
         invec[i] = outvec[i] = FALSE;
   }
 
   if (argv[index] != NIL_PTR) {
         expand(argv[index++], input);
         if (com_fl) complement(input);
         if (argv[index] != NIL_PTR) expand(argv[index], output);
         if (argv[index] != NIL_PTR) map(input, output);
         for (ptr = input; *ptr; ptr++) invec[*ptr] = TRUE;
         for (ptr = output; *ptr; ptr++) outvec[*ptr] = TRUE;
   }
   convert();
   return(0);
 }
 
 void convert()
 {
   short read_chars = 0;
   short c, coded;
   short last = -1;
 
   for (;;) {
         if (in_index == read_chars) {
                 if ((read_chars = read(0, (char *)input, BUFFER_SIZE)) <= 0) {
                         if (write(1, (char *)output, out_index) != out_index)
                                 write(2, "Bad write\n", 10);
                         exit(0);
                 }
                 in_index = 0;
         }
         c = input[in_index++];
         coded = vector[c];
         if (del_fl && invec[c]) continue;
         if (sq_fl && last == coded && outvec[coded]) continue;
         output[out_index++] = last = coded;
         if (out_index == BUFFER_SIZE) {
                 if (write(1, (char *)output, out_index) != out_index) {
                         write(2, "Bad write\n", 10);
                         exit(1);
                 }
                 out_index = 0;
         }
   }
 
   /* NOTREACHED */
 }
 
 void map(string1, string2)
 register unsigned char *string1, *string2;
 {
   unsigned char last;
 
   while (*string1) {
         if (*string2 == '\0')
                 vector[*string1] = last;
         else
                 vector[*string1] = last = *string2++;
         string1++;
   }
 }
 
 static int starts_with(const char *s1, const char *s2)
 {
         while (*s1 && *s1 == *s2)
         {
                 s1++;
                 s2++;
         }
         return *s1 == 0;
 }
 
 /* 
  * character classes from 
  * http://www.opengroup.org/onlinepubs/009695399/utilities/tr.html
  * missing: blank, punct, cntrl, graph, print, space 
  */
 static struct
 {
         const char *keyword;
         char first;
         char last;
 } expand_keywords[] = {
         { "[:alnum:]", 'A', 'Z' },
         { "[:alnum:]", 'a', 'z' },
         { "[:alnum:]", '0', '9' },
         { "[:alpha:]", 'A', 'Z' },
         { "[:alpha:]", 'a', 'z' },
         { "[:digit:]", '0', '9' },
         { "[:lower:]", 'a', 'z' },
         { "[:upper:]", 'A', 'Z' },
         { "[:xdigit:]", '0', '9' },             
         { "[:xdigit:]", 'A', 'F' },
         { "[:xdigit:]", 'a', 'f' }
 };
 
 #define LENGTH(a) ((sizeof((a))) / (sizeof((a)[0])))
 
 void expand(arg, buffer)
 register char *arg;
 register unsigned char *buffer;
 {
   int i, ac, keyword_index;
 
   while (*arg) {
         if (*arg == '\\') {
                 arg++;
                 i = ac = 0;
                 if (*arg >= '0' && *arg <= '7') {
                         do {
                                 ac = (ac << 3) + *arg++ - '0';
                                 i++;
                         } while (i < 4 && *arg >= '0' && *arg <= '7');
                         *buffer++ = ac;
                 } else if (*arg != '\0')
                         *buffer++ = *arg++;
         } else if (*arg == '[') {
                 /* does one of the keywords match? */
                 keyword_index = -1;
                 for (i = 0; i < LENGTH(expand_keywords); i++)
                         if (starts_with(expand_keywords[i].keyword, arg))
                         {
                                 /* we have a match, remember and expand */
                                 keyword_index = i;
                                 ac = expand_keywords[i].first;
                                 while (ac <= expand_keywords[i].last)
                                         *buffer++ = ac++;
                         }
 
                 /* skip keyword if found, otherwise expand range */
                 if (keyword_index >= 0)
                         arg += strlen(expand_keywords[keyword_index].keyword);
                 else
                 {
                         /* expand range */
                         arg++;
                         i = *arg++;
                         if (*arg++ != '-') {
                                 *buffer++ = '[';
                                 arg -= 2;
                                 continue;
                         }
                         ac = *arg++;
                         while (i <= ac) *buffer++ = i++;
                         arg++;          /* Skip ']' */
                 }
         } else
                 *buffer++ = *arg++;
   }
 }
 
 void complement(buffer)
 unsigned char *buffer;
 {
   register unsigned char *ptr;
   register short i, index;
   unsigned char conv[ASCII + 2];
 
   index = 0;
   for (i = 1; i <= ASCII; i++) {
         for (ptr = buffer; *ptr; ptr++)
                 if (*ptr == i) break;
         if (*ptr == '\0') conv[index++] = i & ASCII;
   }
   conv[index] = '\0';
   strcpy((char *)buffer, (char *)conv);
 }
